lib/avro/builder/definition_cache.rb

```diff
@@ -0,0 +1,64 @@
+module Avro
+  module Builder
+
+    # This class is used to cache previously defined schema objects
+    # preventing direct access via the DSL.
+    class DefinitionCache
+      def initialize(builder)
+        @builder = builder
+        @schema_objects = {}
+        @schema_names = Set.new
+      end
+
+      # Cache and schema object by name (for convenience) and fullname.
+      # The schema object is only available by name if the unqualified name is unique.
+      def add_schema_object(object)
+        store_by_name(object) if object.namespace
+        store_by_fullname(object)
+      end
+
+      # Lookup an Avro schema object by name, possibly fully qualified by namespace.
+      def lookup_named_type(key, namespace = nil)
+        key_str = Avro::Name.make_fullname(key.to_s, namespace && namespace.to_s)
+        object = schema_objects[key_str]
+
+        unless object || schema_names.include?(key.to_s)
+          object = builder.import(key)
+        end
+
+        raise DefinitionNotFoundError.new(key) unless object
+        object
+      rescue DefinitionNotFoundError
+        if namespace
+          lookup_named_type(key, nil)
+        else
+          raise
+        end
+      end
+
+      private
+
+      # Schemas are stored by name, provided that the name is unique.
+      # If the unqualified name is ambiguous then it is removed from the cache.
```

Comment:
Will this approach work if there are three ambiguous references? If I'm reading this correctly it will:

1. Add the first reference
2. Remove the first reference when attempting to adding the second reference because it's ambiguous
3. Add the third reference